include/deal.II/lac/la_parallel_vector.templates.h

```diff
@@ -128,27 +129,138 @@ namespace LinearAlgebra
+
+              const bool contiguous_allocation_enabled = false; // TODO
+
+              if (contiguous_allocation_enabled)
+                MPI_Info_set(info, "alloc_shared_noncontig", "false");
+              else
+                MPI_Info_set(info, "alloc_shared_noncontig", "true");
+
+              const std::size_t align_by = 64;
+
+              std::size_t s =
+                ((new_alloc_size * sizeof(Number) + align_by - 1) /
+                 sizeof(Number)) *
+                sizeof(Number);
+
+              // data.memory_constumption_values = s; // TODO
+
+              MPI_Win_allocate_shared(
+                s, sizeof(Number), info, comm_shared, &data_this, win);
+
+              for (unsigned int i = 0; i < size_sm; i++)
+                {
+                  int      disp_unit;
+                  MPI_Aint ssize;
+                  MPI_Win_shared_query(*win, i, &ssize, &disp_unit, &others[i]);
+                }
+
+              Number *ptr_unaligned = others[rank_sm];
+              Number *ptr_aligned   = ptr_unaligned;
+
+              AssertThrow(std::align(align_by,
+                                     new_alloc_size * sizeof(Number),
+                                     reinterpret_cast<void *&>(ptr_aligned),
+                                     s) != nullptr,
+                          ExcNotImplemented());
+
+              unsigned int n_align_local = ptr_aligned - ptr_unaligned;
+              std::vector<unsigned int> n_align_sm(size_sm);
+
+              MPI_Allgather(&n_align_local,
+                            1,
+                            MPI_UNSIGNED,
+                            n_align_sm.data(),
+                            1,
+                            MPI_UNSIGNED,
+                            comm_shared);
+
+              for (unsigned int i = 0; i < size_sm; i++)
+                others[i] += n_align_sm[i];
+
+              std::vector<unsigned int> new_alloc_sizes(size_sm);
+
+              MPI_Allgather(&new_alloc_size,
+                            1,
+                            MPI_UNSIGNED,
+                            new_alloc_sizes.data(),
+                            1,
+                            MPI_UNSIGNED,
+                            comm_shared);
+
+              data.values_sm.resize(size_sm);
+              for (unsigned int i = 0; i < size_sm; i++)
+                data.values_sm[i] =
+                  ArrayView<const Number>(others[i], new_alloc_sizes[i]);
+
+              data.values = {ptr_aligned, [win](Number *) {
+                               MPI_Win_free(win);
+                               delete (win);
+                             }};
+
+#  ifdef DEBUG
+              for (unsigned int i = 0; i < new_alloc_size; ++i)
+                ptr_aligned[i] = 0.0;
+
+              MPI_Barrier(comm_shared);
+
+              Number temp = 0;
+
+              for (const auto &other : data.values_sm)
+                for (const auto &o : other)
+                  temp += o;
```

Comment:
is there an assert missing here?

Reply:
No. In debug mode, I would like to access all data in the SM domain to make sure that everything is allocated correctly. I have added a comment.

Reply:
But you are doing just a write to a double. I would not be surprised if even with O1 this will be optimized away.

Reply:
I have removed the zeroing.